content/source/docs/extend/debugging.html.md

```diff
@@ -0,0 +1,179 @@
+provider developers, though sufficiently advanced users may also be able to use
+it.
+
+There are two available approaches to debugging Terraform providers. We'll talk
+about each of them separately and in turn.
+
+## Log-Based Debugging
+
+Log-based debugging is a method of using logging calls to record what is
+happening in a provider as it happens, and then examining that record to piece
+together what happened after the fact. Usefully, end users can capture logs and
+share them with developers, allowing developers to sometimes debug providers
+without needing access to the environment or configuration. Unfortunately,
+developers need to think ahead about what information they'll need when
+debugging, because the log lines need to be injected into the provider when
+it's built and cannot be inserted into the binary after it has been built. This
+makes it harder to ask new questions when doing log-based debugging, as the log
+line needs to be added and the provider needs to be recompiled before new
+information can be surfaced.
+
+### Inserting Log Lines Into A Provider
+
+Terraform providers all log to `stderr`. Any output to `stderr` will be passed
+over the gRPC protocol to Terraform, which will then write it to its
+destination.
+
+~> **NOTE:** Don't write to `stdout` on any Terraform providers! It can cause
+   problems with our plugin system, which uses `stdout` as a communication
+   channel.
+
+It is recommended to use the standard library `log` package to achieve this,
+using `log.Println` or `log.Printf` or similar functions.
+
+Log lines must be prefixed with the level of the log in square brackets. Valid
+levels are:
+
+* ERROR
+* WARN
+* INFO
+* DEBUG
+* TRACE
+
+A full log line, then, may look like this:
+
+```go
+log.Println("[DEBUG] Something happened!")
+```
+
+### Turning On Logging
+
+Terraform controls which log levels are recorded for all providers and
+Terraform itself using the `TF_LOG` environment variable. Set it to `TRACE` to
+output all log lines, `DEBUG` to output all log lines except `TRACE` lines,
+etc.
+
+~> **NOTE:** prior to v0.15.0 of Terraform, levels besides `TRACE` may not be
+   entirely reliable. See [this
+   PR](https://github.com/hashicorp/terraform/pull/26632) for more details.
+
+By default, log output will be recorded to `stderr` when `TF_LOG` is set. To
+write log output to a file instead, set `TF_LOG_PATH` to a file's path. Log
+output will be appended there. It's worth noting that this file will not be
+truncated before log output is written.
+
+#### Separating Out Provider Logs
+
+Terraform can have some verbose logging and it can be difficult to parse log
+files when both Terraform and providers are writing to the same log file. To
+address this, following the 0.15.0 release of Terraform, the `TF_LOG_CORE`
+environment variable and `TF_LOG_PROVIDER` environment variable can now be set
+to different levels. Set either to `off` to turn off its logging entirely.
```

Comment:
```suggestion
files when both Terraform and providers are writing to the same log file. In
Terraform 0.15.0 and later, you can address this by setting the `TF_LOG_CORE`
and `TF_LOG_PROVIDER` environment variables to different log levels.
Set either to `off` to turn off its logging entirely.
```

- Reducing passive voice.
- Putting version requirement earlier in that sentence.